src/ray/raylet/scheduling/local_resource_manager.h

```diff
@@ -0,0 +1,218 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <gtest/gtest_prod.h>
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "ray/common/task/scheduling_resources.h"
+#include "ray/gcs/gcs_client/accessor.h"
+#include "ray/gcs/gcs_client/gcs_client.h"
+#include "ray/raylet/scheduling/cluster_resource_data.h"
+#include "ray/raylet/scheduling/cluster_resource_scheduler_interface.h"
+#include "ray/raylet/scheduling/fixed_point.h"
+#include "ray/raylet/scheduling/scheduling_ids.h"
+#include "ray/raylet/scheduling/scheduling_policy.h"
+#include "ray/util/logging.h"
+#include "src/ray/protobuf/gcs.pb.h"
+
+namespace ray {
+
+using rpc::HeartbeatTableData;
+
+/// Class manages the resources of the local node.
+class LocalResourceManager {
+ public:
+  /// Constructor initializing the resources associated with the local node.
+  ///
+  /// \param local_node_id: ID of local node,
+  /// \param local_node_resources: The total and the available resources associated
+  /// with the local node.
+  LocalResourceManager(int64_t local_node_id, StringIdMap &resource_name_to_id,
+                       const NodeResources &local_node_resources);
+  LocalResourceManager(
+      int64_t local_node_id, StringIdMap &resource_name_to_id,
+      const absl::flat_hash_map<std::string, double> &local_node_resources,
+      std::function<int64_t(void)> get_used_object_store_memory = nullptr,
+      std::function<bool(void)> get_pull_manager_at_capacity = nullptr);
+
+  /// Add a local resource that is available.
+  ///
+  /// \param resource_name: Resource which we want to update.
+  /// \param resource_total: New capacity of the resource.
+  void AddLocalResourceInstances(const std::string &resource_name,
+                                 const std::vector<FixedPoint> &instances);
+
+  /// Delete a given resource from the local node.
+  ///
+  /// \param resource_name: Resource we want to delete
+  void DeleteLocalResource(const std::string &resource_name);
+
+  /// Check whether the available resources are empty.
+  ///
+  /// \param resource_name: Resource which we want to check.
+  bool IsAvailableResourceEmpty(const std::string &resource_name);
+
+  /// Return local resources.
+  NodeResourceInstances GetLocalResources() const { return local_resources_; }
+
+  /// Allocate local resources to satisfy a given request (resource_request).
+  ///
+  /// \param resource_request: Resources requested by a task.
+  /// \param task_allocation: Local resources allocated to satsify resource_request
+  /// demand.
+  ///
+  /// \return true, if allocation successful. If false, the caller needs to free the
+  /// allocated resources, i.e., task_allocation.
+  bool AllocateTaskResourceInstances(
```

Comment:
What's the diff between `AllocateTaskResourceInstances` vs `AllocateLocalTaskResources`? Should we deprecate one of them?